### PR TITLE
Rubi progress and tests

### PR DESCRIPTION
**Pull request by ashishkg0022**
* integrals
    * Made rubi stable.
    * Added missing utility functions
    * Added large test set

### CODE REVIEW

**Comment by ashishkg0022** [comment]
In original test suite for trinomial products there are 5 files. In sympy format I have kept it here. https://github.com/ashishkg0022/rubi-structure/tree/master/test/trinomial . This contains all trinomial tests from the rubi test-suite.
I have added `test_trinomial_passed`  in this PR . This contains tests which are passed for trinomial_1 . Almost 600 test have passed for trinomial_1 . I started testing trinomial_2 . Then I encountered with issues like 

- `With` Function is not handled properly. It should be parsed correctly.

- Some times `Matchpy` wrongly matches `x` with some integer

- and some small issues.
The most important for now is fixing the `With` issue. Then majority of trinomial_2 should pass.

**Comment by ashishkg0022** [comment]
ping @Upabjojr  @parsoyaarihant

**Comment by ashishkg0022** [comment]
Also  
```python
In [1]: from sympy.abc import a,b,c,x, e, d

In [2]: b*e - c*d
Out[2]: b*e - c*d
```
but in mathematica 
```
In[1]:= b*e - c*d
Out[1]= -cd+be
 ```
This results in difference because there is a function `PosQ` which treats `b*e -c*d` True and `-c*d +b*e` False

**Comment by parsoyaarihant** [comment]
> Some times Matchpy wrongly matches x with some integer
This should be reported to MatchPy developers. Can you share a sample case where this fails?

**Comment by parsoyaarihant** [comment]
Regadring difference in outout of SymPy and Mathematica, Mathematica sorts the expression canonically. See: http://reference.wolfram.com/language/ref/Orderless.html http://reference.wolfram.com/language/ref/Sort.html 

You can use `Sort` in the utility function to address this issue.

**Comment by ashishkg0022** [comment]
> You can use Sort in the utility function to address this issue.

It handles rest of the cases . But specifically the case `b*e -c*d` is not sorted by the `Sort` function too.

**Comment by Upabjojr** [comment]
would you feel like separating the parsetools and the generated code in two separate PRs?

**Comment by ashishkg0022** [comment]
Closing this Pr and opening two new PRs. One including generated rules, constraints, tests . Other including parsetools and utility_functions ?

**Comment by Upabjojr** [comment]
Sorry, ignore my previous message. It's better to stay here.

**Comment by Upabjojr** [comment]
Some files are automatically generated, right?

**Comment by Upabjojr** [comment]
In the file ` sympy/integrals/rubi/parsetools/header.py.txt`, better put a warning that the code to automatically generated, and not to edit.

**Comment by Upabjojr** [comment]
Are you using my repository to generate the code (with the code I have parsed from Mathematica)? Or did you use your own?

**Comment by ashishkg0022** [comment]
Yes

**Comment by ashishkg0022** [comment]
> Are you using my repository to generate the code (with the code I have parsed from Mathematica)? Or did you use your own?

I have just updated last year's parsetools.

**Comment by Upabjojr** [comment]
The file `/usr/people/home/bonazzi/sympy/sympy/integrals/rubi/__init__.py` needs more details.

**Comment by ashishkg0022** [comment]
Your repository contains all parsed rules in a single file. I have parsed them separately. Should I create a repo for that ?

**Comment by Upabjojr** [comment]
> Your repository contains all parsed rules in a single file. I have parsed them separately. Should I create a repo for that ?

We need a way to allow a user to simply launch a command and generate the rules. Like automatic retrieval.

All instructions should be written in `/usr/people/home/bonazzi/sympy/sympy/integrals/rubi/__init__.py`

**Comment by Upabjojr** [comment]
Even the Mathematica commands should be described.

**Comment by ashishkg0022** [comment]
> We need a way to allow a user to simply launch a command and generate the rules. Like automatic retrieval.

Getting the `DownValues` has to be done in steps. I loaded all rules (files of each type) one by one then generated the `DownValues` for each. After we get the downvalues, everything can be generated by one command.

**Comment by Upabjojr** [comment]
can you put that into a script?

**Comment by ashishkg0022** [comment]
script for generating the rule files from downvalues?

**Comment by Upabjojr** [comment]
Yes, also a script for generating the DownValues... let's create a nice documentation with all steps to parse the rules.

**Comment by ashishkg0022** [comment]
Should I create something like `rubi_parsing_guide.md` . 
Also will we be including those script files in sympy or instead of files, I write the code in the `rubi_parsing_guide.md` ?

**Comment by ashishkg0022** [comment]
I have added this https://github.com/ashishkg0022/sympy/blob/5dacc8d6e6df150e44a3d025c42294c0015bc3e1/sympy/integrals/rubi/parsetools/rubi_parsing_guide.md
along with `generate_rules.py` . 

I think it will be good If I make a function which parses the test cases too from mathematica format. We should not rely upon maple format. Should I add it in `rubi_test` submodule?

**Comment by Upabjojr** [comment]
> We should not rely upon maple format. Should I add it in rubi_test submodule?

Yes, add everything that may be useful.

**Comment by ashishkg0022** [comment]
I added `generate_tests.py` for parsing tests and tested test cases of `error_functions` with this parser. It is working fine. I added the steps in `rubi_parsing_guide.md`. Please check this
https://github.com/ashishkg0022/sympy/blob/97db9b310dc90900f6ff005a82c791002711493f/sympy/integrals/rubi/parsetools/rubi_parsing_guide.md

**Comment by Abdullahjavednesar** [comment]
@ashishkg0022 have you implemented all the utility functions? What are you working on next? When do you plan to start with the documentation?

**Comment by ashishkg0022** [comment]
Yes, I have implemented all. There is a missing implementation in `apart`. I have asked on gitter about how should I proceed next

**Comment by Abdullahjavednesar** [comment]
That's good, so what do you plan to do next?

**Comment by ashishkg0022** [comment]
I will add more instructions and documentation in this PR. I want this to merged as soon as possible so that I can move on to update utility functions of the latest version.

**Comment by Abdullahjavednesar** [comment]
@ashishkg0022 for those utility functions which you've recently modified have you added relevant tests for all (which probably was failing before)? Although I've seen your work before looks good to me, but if you are missing anywhere just recheck.

**Comment by ashishkg0022** [comment]
Yes, I think I have added test cases for them.

**Comment by ashishkg0022** [comment]
There are some more things that need to be done to improve utility functions. 
I have opened a separate PR especially for improving utility functions. ( #14956)

**Comment by ashishkg0022** [comment]
I apologise, I have not run pyflakes on it. I was not aware of it (pyflakes). I will run it today and fixed if something is undefined. I will fix it today. Then this can be merged.

**Comment by ashishkg0022** [comment]
Fixed few undefined error using pyflakes.

**Comment by certik** [comment]
@Upabjojr would you mind reviewing this PR please and give it a final +1?

**Comment by Upabjojr** [comment]
@ashishkg0022 did you add the instructions on how to run the tests?

File: sympy/integrals/rubi/__init__.py [hunk 1]
```diff
@@ -24,10 +69,10 @@
...
 the utility functions used in the `ReplacementRule`.
 
-Parsing Rules
-=============
-I have included the parser in /parsetools folder. The parser takes
-`FullForm[DownValues[]]` of Mathematica rules as input.
+Parsing Rules and Tests
+=======================
+Code for parsing rule and tests are included in sympy.
+They have been properly explained with steps in `rubi_parsing_guide.md` in parsetools.
```
Comment (Upabjojr) [inline sympy/integrals/rubi/__init__.py]: full path of `rubi_parsing_guide.md` is better

File: sympy/utilities/runtests.py [hunk 2]
```diff
@@ -671,6 +671,7 @@ def _doctest(*paths, **kwargs):
         "sympy/physics/unitsystems.py", # raises deprecation warning
         "sympy/parsing/latex/_antlr/latexlexer.py", # generated code
         "sympy/parsing/latex/_antlr/latexparser.py", # generated code
+        "sympy/integrals/rubi/rubi.py"
```
Comment (Upabjojr) [inline sympy/utilities/runtests.py]: are we blacklisting the doctests as well?

Comment (ashishkg0022) [reply sympy/utilities/runtests.py]: only for `rubi.py`, not for utility_functions. 
Actually, the loading time is around 10 - 11 minutes. So, travis fails. If nothing happens in 10 minutes, travis fails. So I have blacklisted it.

**Comment by Upabjojr** [comment]
I guess the master branch should be merged into this PR.

**Comment by ashishkg0022** [comment]
> I guess the master branch should be merged into this PR.

I have merged the master branch

**Comment by ashishkg0022** [comment]
The travis failed due to some network error.

**Comment by Upabjojr** [comment]
Some of the rubi tests keep failing. For example, `test_1_3.py`, is it supposed to be like that?

**Comment by ashishkg0022** [comment]
`test_1_3.py` is from last year. I have not run it due to a large time taken by it. 
What is the status of newly added test files? Do a lot of them fail?

**Comment by Upabjojr** [comment]
> Do a lot of them fail?

I didn't test them all so far.

**Comment by ashishkg0022** [comment]
Can you please run them all? I have tested them in part. I have not run them all together. Like, when I tested exponentials they passed. But after a week I made some changes in trigo rules and utility functions. So it might be possible very few might fail because of some changes done later. For any new change, it's not possible to run all tests again.

**Comment by asmeurer** [comment]
Regarding the release notes, I accidentally added `rubi` as a submodule, but since it's been moved to `integrals.rubi`, the header should just be "integrals" (I've fixed it). c.f. https://github.com/sympy/sympy-bot/pull/16. If there end up being a lot of integrals release notes we can split them up into different subsubmodules like integrals.risch, integrals.rubi, and so on.

**Comment by Upabjojr** [comment]
For the rubi_test results:
https://gist.github.com/Upabjojr/02cf3005ec995c7401418624a4bc11c7

**Comment by ashishkg0022** [comment]
I saw it. Tests which I have run recently like hyperbolic are passing but a few of trinomials and miscellaneous algebra might be failing. Is there some way I can get a list of all expressions in the test set which are failing?

**Comment by Upabjojr** [comment]
You have the line numbers. Some tests are actually still running.

**Pull request merged**